File: labs_app/inventory_management/report/my_dividend_report/my_dividend_report.js
// Copyright (c) 2025, Your Company and contributors
// For license information, please see license.txt

frappe.query_reports["My Dividend Report"] = {
	"filters": [
		{
			"fieldname": "fiscal_year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year"
		},
        {
			"fieldname": "member_name",
			"label": __("Member Name"),
			"fieldtype": "Link",
			"options": "Investment Member",
		},
		{
			"fieldname": "dividend_rate",
			"label": __("Interest on Shares (%)"),
			"fieldtype": "Float"
		},
		{
			"fieldname": "interest_rate",
			"label": __("Interest on Deposit (%)"),
			"fieldtype": "Float"
		},
        {
			"fieldname": "withholding_tax_rate",
			"label": __("Withholding Tax Rate (%)"),
			"fieldtype": "Float"
		},
		{
			"fieldname": "processing_fee",
			"label": __("Processing Fee"),
			"fieldtype": "Currency"
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Right align currency columns
		if (column.fieldtype === "Currency" || column.fieldname.includes("deposit") || 
			column.fieldname.includes("capital") || column.fieldname.includes("dividend") || 
			column.fieldname.includes("interest") || column.fieldname.includes("fee") || 
			column.fieldname.includes("tax")) {
			if (value && value !== "0.00") {
				value = '<div style="text-align: right">' + value + '</div>';
			}
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Add Export to CSV button
		report.page.add_inner_button(__('Export to CSV'), function() {
			const data = report.data;
			const columns = report.columns;
			
			if (!data || data.length === 0) {
				frappe.msgprint(__('No data to export'));
				return;
			}
			
			// Create CSV content
			let csv = [];
			
			// Add title row
			csv.push(['COMPUTATION OF DIVIDEND & INTEREST FY 2024']);
			csv.push([]);
			
			// Add headers
			let headers = columns.map(col => col.label || col.fieldname);
			csv.push(headers);
			
			// Add data rows
			data.forEach((row, index) => {
				let csv_row = columns.map(col => {
					let value = row[col.fieldname];
					if (value === null || value === undefined) {
						return '';
					}
					// Handle strings with commas
					if (typeof value === 'string' && value.includes(',')) {
						return '"' + value + '"';
					}
					return value;
				});
				csv.push(csv_row);
			});
			
			// Convert to CSV string
			let csv_content = csv.map(row => row.join(',')).join('\n');
			
			// Create download
			let blob = new Blob([csv_content], { type: 'text/csv;charset=utf-8;' });
			let link = document.createElement("a");
			let url = URL.createObjectURL(blob);
			link.setAttribute("href", url);
			link.setAttribute("download", "dividend_computation_" + frappe.datetime.now_date() + ".csv");
			link.style.visibility = 'hidden';
			document.body.appendChild(link);
			link.click();
			document.body.removeChild(link);
			
			frappe.msgprint(__('CSV file downloaded successfully'));
		}, __('Actions'));
		
		// Add Print button
		report.page.add_inner_button(__('Print'), function() {
			window.print();
		}, __('Actions'));
	}
};

File: labs_app/inventory_management/report/my_dividend_report/my_dividend_report.py
# Copyright (c) 2025, Your Company and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
	"""Main execution function for the report"""
	columns = get_columns()
	data = get_data(filters)
	return columns, data


def get_columns():
	"""Define all report columns"""
	return [
		{
			"fieldname": "member_number",
			"label": _("No"),
			"fieldtype": "Data",
			"width": 120
		},
		{
			"fieldname": "member_name",
			"label": _("Name"),
			"fieldtype": "Data",
			"width": 200
		},
		{
			"fieldname": "current_deposit",
			"label": _("Current Deposit"),
			"fieldtype": "Currency",
			"width": 140
		},
		{
			"fieldname": "share_capital",
			"label": _("Shares Capital"),
			"fieldtype": "Currency",
			"width": 130
		},
		{
			"fieldname": "qualifying_current_deposits",
			"label": _("Qualifying Current Deposits"),
			"fieldtype": "Currency",
			"width": 160
		},
		{
			"fieldname": "qualifying_shares_retained",
			"label": _("Qualifying Shares Retained"),
			"fieldtype": "Currency",
			"width": 160
		},
		{
			"fieldname": "explanation",
			"label": _("Explanation"),
			"fieldtype": "Data",
			"width": 300
		},
		{
			"fieldname": "gross_dividends",
			"label": _("Gross Dividends. 7% of share capital"),
			"fieldtype": "Currency",
			"width": 150
		},
		{
			"fieldname": "gross_interest_on_deposit",
			"label": _("Gross Interest on Deposit. 6%"),
			"fieldtype": "Currency",
			"width": 150
		},
		{
			"fieldname": "processing_fee",
			"label": _("Processing Fee"),
			"fieldtype": "Currency",
			"width": 120
		},
		{
			"fieldname": "withholding_tax",
			"label": _("Withholding tax 5%"),
			"fieldtype": "Currency",
			"width": 140
		},
		{
			"fieldname": "net_dividends",
			"label": _("Net Dividends"),
			"fieldtype": "Currency",
			"width": 130
		}
	]


def get_data(filters=None):
	"""Fetch and calculate dividend data"""
	
	if not filters:
		filters = frappe._dict()
	
	# Get filter values with defaults
	dividend_rate = flt(filters.get("dividend_rate")) 
	interest_rate = flt(filters.get("interest_rate")) 
	processing_fee = flt(filters.get("processing_fee"))
	withholding_tax_rate = flt(filters.get("withholding_tax_rate"))
	
	# Build WHERE conditions
	conditions = ["1=1"]
	
	# if filters.get("member_name"):
	# 	conditions.append(f"member_name = %(member_name)s")
	
	
	where_clause = " AND ".join(conditions)
	
	# Fetch member data from Dividend Calculator Detail
	query = f"""
		SELECT 
			member_number,
			member_name,
			COALESCE(current_deposit, 0) as current_deposit,
			COALESCE(share_capital, 0) as share_capital,
			COALESCE(qualifying_current_deposits, 0) as qualifying_current_deposits,
			COALESCE(qualifying_shares_retained, 0) as qualifying_shares_retained,
			qualifying_deposit_explanation,
			qualifying_share_explanation
		FROM 
			`tabDividend Calculator Detail`
		WHERE
			{where_clause}
		ORDER BY 
			member_number
	"""
	
	members = frappe.db.sql(query, filters, as_dict=1)
	
	# Calculate dividends for each member
	data = []
	for member in members:
		# Get values
		share_capital = flt(member.share_capital)
		qualifying_deposits = flt(member.qualifying_current_deposits)
		
		# Calculate gross dividends (based on share capital)
		gross_dividends = share_capital * dividend_rate
		
		# Calculate gross interest (based on qualifying deposits)
		gross_interest = qualifying_deposits * interest_rate
		
		# Calculate total gross income
		total_gross = gross_dividends + gross_interest
		
		# Calculate withholding tax (5% of total gross)
		withholding_tax = total_gross * withholding_tax_rate
		
		# Calculate net dividends
		net_dividends = total_gross - withholding_tax - processing_fee
		
		# Prepare combined explanation
		explanation = get_combined_explanation(
			member.current_deposit, 
			member.qualifying_current_deposits,
			member.share_capital,
			member.qualifying_shares_retained,
			member.get("qualifying_deposit_explanation"),
			member.get("qualifying_share_explanation")
		)
		
		# Append row
		data.append({
			"member_number": member.member_number,
			"member_name": member.member_name,
			"current_deposit": flt(member.current_deposit, 2),
			"share_capital": flt(share_capital, 2),
			"qualifying_current_deposits": flt(qualifying_deposits, 2),
			"qualifying_shares_retained": flt(member.qualifying_shares_retained, 2),
			"explanation": explanation,
			"gross_dividends": flt(gross_dividends, 2),
			"gross_interest_on_deposit": flt(gross_interest, 2),
			"processing_fee": flt(processing_fee, 2),
			"withholding_tax": flt(withholding_tax, 2),
			"net_dividends": flt(net_dividends, 2)
		})
	
	return data


def get_combined_explanation(current_deposit, qualifying_deposit, share_capital, qualifying_shares, 
							  deposit_explanation_text=None, share_explanation_text=None):
	"""Generate combined explanation for both qualifying deposits and shares"""
	
	# If custom explanations exist in the doctype, use them
	if deposit_explanation_text and share_explanation_text:
		return f"{deposit_explanation_text}; {share_explanation_text}"
	
	# Otherwise, generate automatic explanation
	explanations = []
	
	# Deposit explanation
	current = flt(current_deposit)
	qualifying_dep = flt(qualifying_deposit)
	
	if current == qualifying_dep and current > 0:
		explanations.append("Full deposit qualifies")
	elif qualifying_dep == 0 and current > 0:
		explanations.append("No qualifying deposits")
	elif qualifying_dep > 0:
		explanations.append(f"Deposit: {qualifying_dep:,.2f} of {current:,.2f} qualifies")
	
	# Share explanation
	capital = flt(share_capital)
	qualifying_shr = flt(qualifying_shares)
	
	if capital == qualifying_shr and capital > 0:
		explanations.append("All shares qualify")
	elif qualifying_shr == 0 and capital > 0:
		explanations.append("No qualifying shares")
	elif qualifying_shr > 0:
		explanations.append(f"Shares: {qualifying_shr:,.2f} of {capital:,.2f} qualify")
	
	return "; ".join(explanations) if explanations else "No qualifying amounts"